dipy/tracking/local/localtrack.pyx

```diff
@@ -208,3 +199,301 @@ cdef int _local_tracker(
+        Temporary array for directions followed by particles.
+    particle_weights : array, float, 2d, (2, particle_count)
+        Temporary array for the weights of particles.
+    particle_states : array, float, (2, particle_count, 2)
+        Temporary array for the states of particles.
+
+    Returns
+    -------
+    end : int
+        Length of the tracked streamline
+    tissue_class : TissueClass
+        Ending state of the streamlines as determined by the TissueClassifier.
+
+    """
+    cdef:
+        int i
+        TissueClass tissue_class
+        double dir[3]
+        double vs[3]
+        double seed[3]
+
+    if (seed_pos.shape[0] != 3 or first_step.shape[0] != 3 or
+            voxel_size.shape[0] != 3 or streamline.shape[1] != 3):
+        raise ValueError('Invalid input parameter dimensions.')
+
+    for i in range(3):
+        dir[i] = first_step[i]
+        vs[i] = voxel_size[i]
+        seed[i] = seed_pos[i]
+
+    i = _pft_tracker(dg, tc, seed, dir, vs, streamline,
+                     directions, step_size, &tissue_class, pft_nbr_back_steps,
+                     pft_max_steps, pft_max_trials, particle_count,
+                     particle_paths, particle_dirs, particle_weights,
+                     particle_states)
+    return i, tissue_class
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef _pft_tracker(DirectionGetter dg,
+                  ConstrainedTissueClassifier tc,
+                  double* seed,
+                  double* dir,
+                  double* voxel_size,
+                  np.float_t[:, :] streamline,
+                  np.float_t[:, :] directions,
+                  double step_size,
+                  TissueClass * tissue_class,
+                  int pft_nbr_back_steps,
+                  int pft_max_steps,
+                  int pft_max_trials,
+                  int particle_count,
+                  np.float_t[:, :, :, :] particle_paths,
+                  np.float_t[:, :, :, :] particle_dirs,
+                  np.float_t[:, :] particle_weights,
+                  np.int_t[:, :, :] particle_states):
+    cdef:
+        int i, pft_trial, pft_streamline_i, pft_nbr_steps, strl_array_len
+        double point[3]
+        double voxdir[3]
+        void (*step)(double* , double*, double) nogil
+
+    copypoint(seed, point)
+    copypoint(seed, &streamline[0,0])
+    copypoint(dir, &directions[0, 0])
+
+    tissue_class[0] = TRACKPOINT
+    pft_trial = 0
+    i = 0
+    strl_array_len = streamline.shape[0]
+    while i < strl_array_len - 1:
+        if dg.get_direction_c(point, dir):
+            # no valid diffusion direction to follow
+            tissue_class[0] = INVALIDPOINT
+        else:
+            for j in range(3):
+                voxdir[j] = dir[j] / voxel_size[j]
+            i += 1
+            fixed_step(point, voxdir, step_size)
+            copypoint(point, &streamline[i, 0])
+            copypoint(dir, &directions[i, 0])
```

Comment:
Why are you keeping all directions? You don't seem to be using them in `ParticleFilteringTracking`, or anywhere else.

Reply:
Ah, ok, you are putting some of them in `particle_dirs`, so keeping them all is simpler.

Reply:
Dammit, sorry, I answer too fast then I see something else :) It seems you are filling `particle_dirs` using only the last direction, so I still don't understand why you are keeping all directions.

    for p in range(particle_count):
        copypoint(&directions[streamline_i, 0], &particle_dirs[0, p, 0, 0])

Reply:
I need the direction for the backtracking step. Before starting PFT, we backtrack a few steps (`pft_nbr_back_steps`, e.g. 2mm), then we start the tracking with the particle filter. At this time, we need to know the direction followed to reach the current position. 

Because of the backtracking, streamline segments obtained with PFT can overlap, thus I need to save the direction of each of the streamline point. Also, the backtracking distance is not always the same (e.g. shorter if the streamline is less than 2mm).